readme.txt
=== WP Favorite Posts ===
Contributors: Huseyin Berberoglu
Donate link: http://nxsn.com.com/my-projects/wp-favorite-posts-plugin
Tags: favorite posts, favorite, favourite, posts, favorites,
wp-favorite-posts, reading list, post list, post lists, lists
Requires at least: 2.0.2
Tested up to: 3.0.3
Stable tag: 1.5.5

Allows visitors to add favorite posts. This plugin use cookies for saving data so
unregistered users can favorite a post.

== Description ==

Allows users to add favorite posts. This plugin use cookies and database for
saving data.

- If a user logged in then favorites data will saved in database instead of cookies.
- If user not logged in data will saved in cookies.

You can choose "only registered users can favorite a post" option, if you want.
Also there is a widget named "Most Favorited Posts". And you can use this template
tag for listing most favorited posts;

<h2>Most Favorited Posts</h2>
<?php wpfp_list_most_favorited(5); ?>

If you use WP Super Cache you must add page (which you show favorites) URI to "Accepted Filenames &
Rejected URIs".

See [more details about
plugin](http://nxsn.com/my-projects/wp-favorite-posts/)

== Installation ==

1. Unzip into your `/wp-content/plugins/` directory.
1. Activate the plugin through the 'Plugins' menu in WordPress
1. Place `<?php if (function_exists('wpfp_link')) { wpfp_link(); } ?>` in your
single.php or page.php template file. Then favorite this post link will appear in all posts.
1. OR if you DO NOT want the favorite link to appear in every post/page, DO NOT
use the code above. Just type in [wpfp-link] into the selected post/page
content and it will embed the print link into that post/page only.
1. Create a page e.g. "Your Favorites" and insert `{{wp-favorite-posts}}`
text into content section. This page will contain users favorite posts.
1. That's it :)

== Screenshots ==
1. Options
2. Label Settings
3. Advanced Settings

== Changelog ==
= 1.5.5 =
2011-11-23
	* BUGFIX: related with 1.5.4
	
= 1.5.4 =
2011-11-23
	* BUGFIX: Thanks to @spectrus http://wordpress.org/support/topic/plugin-wp-favorite-posts-major-bug-getting-user-favourites?replies=1#post-2465141
	
= 1.5.3 =
2011-11-23
	* NEW FEATURE: Added don't load js/css file option. If you don't 
	want to load wp favorite post's js/css file, use these options. 
	
= 1.5.2 =
2010-12-27
	* FIX: Change the user's favorites wigdet title bug fixed.

= 1.5.1 =
2010-06-04
	* Little JS fix. Thanks to Heather Wallace.

= 1.4.3.2 =
2010-04-23
	* Show only published posts on most favorited posts widget

= 1.4.3 =
2010-04-13
	* Admin can write html codes to label settings (on admin page)
	* Added "wpfp_link_html" and "wpfp_remove_favorite_link" filters.

= 1.4.3 =
2010-04-09
	* Fix: same remove link for all posts on index
	* better wpfp.js: remove li which on favorites page.

= 1.4.1 =
2010-04-05
	* code refactor, add do_action for add and remove to list (#225575)
	* update admin page

2010-03-29
	* Fix plugin path, fixes image loading problems (#222902)

= 1.4 =
2010-03-20
	* Override page template if wpfp-page-template.php exists on template
	directory.
	* Add [wp-favorite-posts] shortcode. Use shortcode instead of
	{{wp-favorite-posts}}

= 1.3.5 =
2010-03-17
	* Fix meta key issue.

= 1.3.4 =
2010-03-16
	* Fix wpfp-span issue

= 1.3.3 =
2010-03-16
	* Fixed regression: if javascript doesn't work change to non-ajax mode.
	This fixes 'only text' pages. 

2010-02-15
	* If user logged in don't show cookie warning at "your favorites" page.
	* Added "show remove link" and "show add link" options. 
	You can show remove link when someone add a favorite.
	Similary You can show add link when someone remove a favorite.

= 1.3.1 =
2009-06-10
	* Added Before Link Image feature.
	* Refactor code and imporve DRY
	* Fixed bug: Most favorite list's string sorting problem 2 > 11
	* Fixed bug: Clear link not updating post's favorited count

= 1.3 =
2009-05-31
	* Fixed bug: Plugin was working wrong in pages with links includes # character.
	* Added template tag for Most Favorite Posts. There was already a widget for this.

= 1.2.1 =
2009-05-14
	* Added "Most Favorited Posts" widget.

2009-04-30
	* Added "buy me a beer" section to admin page.
	* Added favorite statistic feature.

= 1.2 =
2009-04-26
	* Added database integration. 
		- If a user logged in then favorites data will save to database instead of cookies.
		- If user not logged in data will save to cookies.
	* Added "only registered users can favorite" option.

= 1.1.7 =
2009-03-10
	* Fixed duplicate loading image problem
	* Added [wpfp-link] feature;
		You can show favorite link only in preferred posts with writing
		[wpfp-link] to the post content.

= 1.1.6 =
2009-03-05
	* Fixed ajax problem.

= 1.1.5 =
2009-03-02
	* Added rel="nofollow" to links.
	* Favorite posts title language problem solved.
	* ajax.js file renamed to wpfp.js

2009-03-02
	* Use XHTML valid links.
	* Use class instead of id for html elements.
	* Use more unique function names. All functions starts with wpfp.

= 1.2.1 =
2009-05-14
	* Added "Most Favorited Posts" widget.

2009-04-30
	* Added "buy me a beer" section to admin page.
	* Added favorite statistic feature.

= 1.2 =
2009-04-26
	* Added database integration. 
		- If a user logged in then favorites data will save to database instead of cookies.
		- If user not logged in data will save to cookies.
	* Added "only registered users can favorite" option.

= 1.1.7 =
2009-03-10
	* Fixed duplicate loading image problem
	* Added [wpfp-link] feature;
		You can show favorite link only in preferred posts with writing
		[wpfp-link] to the post content.

= 1.1.6 =
2009-03-05
	* Fixed ajax problem.

= 1.1.5 =
2009-03-02
	* Added rel="nofollow" to links.
	* Favorite posts title language problem solved.
	* ajax.js file renamed to wpfp.js

2009-03-02
	* Use XHTML valid links.
	* Use class instead of id for html elements.
	* Use more unique function names. All functions starts with wpfp.

= 1.1.4 =
2009-02-24
	* Use permalinks favorite links.
	* Users can remove a single favorite post from favorite posts page.

= 1.1.3 =
2009-02-24
	* Clear Favorites now works.
	* Favorite list is xhtml valid now.

= 1.1.2 =
2009-02-24
	* Save button fixed which in options page.

= 1.1.1 =
2009-02-24
	* Cleared wrong code.

= 1.1 =
2009-02-24
	* Fixed permalink problem. Now works with all types of permalink.

= 1.0 =
2009-02-23
	* First Release of WP-Favorite Posts


ChangeLog.txt
------------------
WP Favorite Posts - ChangeLog
http://wordpress.org/extend/plugins/wp-favorite-posts/
http://nxsn.com
------------------
Version 1.5.3
-----------
2011-11-23 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* NEW FEATURE: Added don't load js/css file option. If you don't 
	want to load wp favorite post's js/css file, use these options. 
	
Version 1.5.2
-----------
2010-12-27 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* FIX: Change the user's favorites wigdet title bug fixed.

Version 1.5.1
-----------
2010-06-04 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Little JS fix. Thanks to Heather Wallace.

Version 1.4.3.2
-----------
2010-04-23 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Show only published posts on most favorited posts widget

Version 1.4.3
-----------
2010-04-13 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Admin can write html codes to label settings (on admin page)
	* Added "wpfp_link_html" and "wpfp_remove_favorite_link" filters.

Version 1.4.3
-----------
2010-04-09 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fix: same remove link for all posts on index
	* better wpfp.js: remove li which on favorites page.

Version 1.4.1
-----------
2010-04-05 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* code refactor, add do_action for add and remove to list (#225575)
	* update admin page

2010-03-29 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fix plugin path, fixes image loading problems (#222902)

Version 1.4
-----------
2010-03-20 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Override page template if wpfp-page-template.php exists on template
	directory.
	* Add [wp-favorite-posts] shortcode. Use shortcode instead of
	{{wp-favorite-posts}}

Version 1.3.5
-----------
2010-03-17 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fix meta key issue.

Version 1.3.4
-----------
2010-03-16 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fix wpfp-span issue

Version 1.3.3
-----------
2010-03-16 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fixed regression: if javascript doesn't work change to non-ajax mode.
	This fixes 'only text' pages. 

2010-02-15 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* If user logged in don't show cookie warning at "your favorites" page.
	* Added "show remove link" and "show add link" options. 
	You can show remove link when someone add a favorite.
	Similary You can show add link when someone remove a favorite.

Version 1.3.1
-----------
2009-06-10 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Added Before Link Image feature.
	* Refactor code and imporve DRY
	* Fixed bug: Most favorite list's string sorting problem 2 > 11
	* Fixed bug: Clear link not updating post's favorited count

Version 1.3
-----------
2009-05-31 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fixed bug: Plugin was working wrong in pages with links includes # character.
	* Added template tag for Most Favorite Posts. There was already a widget for this.

Version 1.2.1
-----------
2009-05-14 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Added "Most Favorited Posts" widget.

2009-04-30 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Added "buy me a beer" section to admin page.
	* Added favorite statistic feature.

Version 1.2
-----------
2009-04-26 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Added database integration. 
		- If a user logged in then favorites data will save to database instead of cookies.
		- If user not logged in data will save to cookies.
	* Added "only registered users can favorite" option.

Version 1.1.7
-----------
2009-03-10 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fixed duplicate loading image problem
	* Added [wpfp-link] feature;
		You can show favorite link only in preferred posts with writing
		[wpfp-link] to the post content.

Version 1.1.6
-----------
2009-03-05 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fixed ajax problem.

Version 1.1.5
-----------
2009-03-02 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Added rel="nofollow" to links.
	* Favorite posts title language problem solved.
	* ajax.js file renamed to wpfp.js

2009-03-02 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Use XHTML valid links.
	* Use class instead of id for html elements.
	* Use more unique function names. All functions starts with wpfp.

Version 1.1.4
-----------
Version 1.2.1
-----------
2009-05-14 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Added "Most Favorited Posts" widget.

2009-04-30 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Added "buy me a beer" section to admin page.
	* Added favorite statistic feature.

Version 1.2
-----------
2009-04-26 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Added database integration. 
		- If a user logged in then favorites data will save to database instead of cookies.
		- If user not logged in data will save to cookies.
	* Added "only registered users can favorite" option.

Version 1.1.7
-----------
2009-03-10 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fixed duplicate loading image problem
	* Added [wpfp-link] feature;
		You can show favorite link only in preferred posts with writing
		[wpfp-link] to the post content.

Version 1.1.6
-----------
2009-03-05 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Fixed ajax problem.

Version 1.1.5
-----------
2009-03-02 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Added rel="nofollow" to links.
	* Favorite posts title language problem solved.
	* ajax.js file renamed to wpfp.js

2009-03-02 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Use XHTML valid links.
	* Use class instead of id for html elements.
	* Use more unique function names. All functions starts with wpfp.

Version 1.1.4
-----------
2009-02-24 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Use permalinks favorite links.
	* Users can remove a single favorite post from favorite posts page.

Version 1.1.3
-----------
2009-02-24 Hüseyin Berberoğlu <hberberoglu@gmail.com>
	* Clear Favorites now works.
	* Favorite list is xhtml valid now.

Version 1.1.2
-----------
2009-02-24 Hüseyin Berberoğlu <hberberoglu@gmail.com>

wp-favorite-posts.php
<?php
/*
Plugin Name: WP Favorite Posts
Plugin URI: http://nxsn.com/my-projects/wp-favorite-posts-plugin/
Description: Allows users to add favorite posts. This plugin use cookies for saving data so unregistered users can favorite a post. Put <code>&lt;?php wpfp_link(); ?&gt;</code> where ever you want on a single post. Then create a page which includes that text : <code>{{wp-favorite-posts}}</code> That's it!
Version: 1.5.5
Author: Huseyin Berberoglu
Author URI: http://nxsn.com

*/


/*
    Copyright (c) 2009 Hüseyin Berberoğlu (hberberoglu@gmail.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

define('WPFP_PATH', WP_PLUGIN_URL . '/wp-favorite-posts');
define('WPFP_META_KEY', "wpfp_favorites");
define('WPFP_USER_OPTION_KEY', "wpfp_useroptions");
define('WPFP_COOKIE_KEY', "wp-favorite-posts");

$ajax_mode = 1;

function wp_favorite_posts() {
    $wpfp_options = wpfp_get_options();

    if (isset($_REQUEST['wpfpaction'])):
        global $ajax_mode;
        $ajax_mode = $_REQUEST['ajax'];
        if ($_REQUEST['wpfpaction'] == 'add') {
            wpfp_add_favorite();
        } else if ($_REQUEST['wpfpaction'] == 'remove') {
            wpfp_remove_favorite();
        } else if ($_REQUEST['wpfpaction'] == 'clear') {
            if (wpfp_clear_favorites()) wpfp_die_or_go(wpfp_get_option('cleared'));
            else wpfp_die_or_go("ERROR");
        }
    endif;
}
add_action('template_redirect', 'wp_favorite_posts');

function wpfp_add_favorite($post_id = "") {
    $wpfp_options = wpfp_get_options();
    if ( empty($post_id) ) $post_id = $_REQUEST['postid'];
    if (wpfp_get_option('opt_only_registered') && !is_user_logged_in() )
        wpfp_die_or_go(wpfp_get_option('text_only_registered') );

    if (wpfp_do_add_to_list($post_id)) {
        // added, now?
        do_action('wpfp_after_add', $post_id);
        if (wpfp_get_option('statics')) wpfp_update_post_meta($post_id, 1);
        if (wpfp_get_option('added') == 'show remove link') {
            $str = wpfp_link(1, "remove", 0, array( 'post_id' => $post_id ) );
            wpfp_die_or_go($str);
        } else {
            wpfp_die_or_go(wpfp_get_option('added'));
        }
    }
}
function wpfp_do_add_to_list($post_id) {
    if (wpfp_check_favorited($post_id))
        return true;
    if (is_user_logged_in()) {
        return wpfp_add_to_usermeta($post_id);
    } else {
        return wpfp_set_cookie($post_id, "added");
    }
}

function wpfp_remove_favorite($post_id = "") {
    $wpfp_options = wpfp_get_options();
    if (empty($post_id)) $post_id = $_REQUEST['postid'];
    if (wpfp_do_remove_favorite($post_id)) {
        // removed, now?
        do_action('wpfp_after_remove', $post_id);
        if (wpfp_get_option('statics')) wpfp_update_post_meta($post_id, -1);
        if (wpfp_get_option('removed') == 'show add link') {
            if ($_REQUEST['page']==1):
                $str = '';
            else:
                $str = wpfp_link(1, "add", 0, array( 'post_id' => $post_id ) );
            endif;
            wpfp_die_or_go($str);
        } else {
            wpfp_die_or_go(wpfp_get_option('removed'));
        }
    }
    else return false;;
}

function wpfp_die_or_go($str) {
    global $ajax_mode;
    if ($ajax_mode):
        die($str);
    else:
        wp_redirect($_SERVER['HTTP_REFERER']);
    endif;
}

function wpfp_add_to_usermeta($post_id) {
    $wpfp_favorites = array();
    $wpfp_favorites = wpfp_get_user_meta();
    $wpfp_favorites[] = $post_id;
    wpfp_update_user_meta($wpfp_favorites);
    return true;
}

function wpfp_check_favorited($cid) {
    if (is_user_logged_in()) {
        $favorite_post_ids = wpfp_get_user_meta();
        if ($favorite_post_ids)
            foreach ($favorite_post_ids as $fpost_id)
                if ($fpost_id == $cid) return true;
	} else {
	    if (wpfp_get_cookie()):
	        foreach (wpfp_get_cookie() as $fpost_id => $val)
	            if ($fpost_id == $cid) return true;
	    endif;
	}
    return false;
}

function wpfp_link( $return = 0, $action = "", $show_span = 1, $args = array() ) {
    global $post;
    $post_id = $post->ID;
    extract($args);
    if ($show_span)
        $str = "<span class='wpfp-span'>";
    $str .= wpfp_before_link_img();
    $str .= wpfp_loading_img();
    $wpfp_options = wpfp_get_options();
    if ($action == "remove"):
        $str .= wpfp_link_html($post_id, wpfp_get_option('remove_favorite'), "remove");
    elseif ($action == "add"):
        $str .= wpfp_link_html($post_id, wpfp_get_option('add_favorite'), "add");
    elseif (wpfp_check_favorited($post_id)):
        $str .= wpfp_link_html($post_id, wpfp_get_option('remove_favorite'), "remove");
    else:
        $str .= wpfp_link_html($post_id, wpfp_get_option('add_favorite'), "add");
    endif;
    if ($show_span)
        $str .= "</span>";
    if ($return) { return $str; } else { echo $str; }
}

function wpfp_link_html($post_id, $opt, $action) {
    $link = "<a class='wpfp-link' href='?wpfpaction=".$action."&amp;postid=". $post_id . "' title='". $opt ."' rel='nofollow'>". $opt ."</a>";
    $link = apply_filters( 'wpfp_link_html', $link );
    return $link;
}

function wpfp_get_users_favorites($user = "") {
    $favorite_post_ids = array();

    if (!empty($user)):
        return wpfp_get_user_meta($user);
    endif;

    # collect favorites from cookie and if user is logged in from database.
    if (is_user_logged_in()):
        $favorite_post_ids = wpfp_get_user_meta();
	else:
	    if (wpfp_get_cookie()):
	        foreach (wpfp_get_cookie() as $post_id => $post_title) {
	            array_push($favorite_post_ids, $post_id);
	        }
	    endif;
	endif;
    return $favorite_post_ids;
}

function wpfp_list_favorite_posts( $args = array() ) {
    $user = $_REQUEST['user'];
    extract($args);
    $wpfp_options = wpfp_get_options();
    if (!empty($user)):
        if (!wpfp_is_user_favlist_public($user)):
            $favorite_post_ids = wpfp_get_users_favorites($user);
        endif;
    else:
        $favorite_post_ids = wpfp_get_users_favorites();
    endif;

    if (@file_exists(TEMPLATEPATH.'/wpfp-page-template.php')):
        include(TEMPLATEPATH.'/wpfp-page-template.php');
    else:
        include("wpfp-page-template.php");
    endif;
}

function wpfp_list_most_favorited($limit=5) {
    global $wpdb;
    $query = "SELECT post_id, meta_value, post_status FROM $wpdb->postmeta";
    $query .= " LEFT JOIN $wpdb->posts ON post_id=$wpdb->posts.ID";
    $query .= " WHERE post_status='publish' AND meta_key='".WPFP_META_KEY."' AND meta_value > 0 ORDER BY ROUND(meta_value) DESC LIMIT 0, $limit";
    $results = $wpdb->get_results($query);
    if ($results) {
        echo "<ul>";
        foreach ($results as $o):
            $p = get_post($o->post_id);
            echo "<li>";
            echo "<a href='".get_permalink($o->post_id)."' title='". $p->post_title ."'>" . $p->post_title . "</a> ($o->meta_value)";
            echo "</li>";
        endforeach;
        echo "</ul>";
    }
}

include("wpfp-widgets.php");

function wpfp_loading_img() {
    return "<img src='".WPFP_PATH."/img/loading.gif' alt='Loading' title='Loading' class='wpfp-hide wpfp-img' />";
}

function wpfp_before_link_img() {
    $options = wpfp_get_options();
    $option = $options['before_image'];
    if ($option == '') {
        return "";
    } else if ($option == 'custom') {
        return "<img src='" . $options['custom_before_image'] . "' alt='Favorite' title='Favorite' class='wpfp-img' />";
    } else {
        return "<img src='". WPFP_PATH . "/img/" . $option . "' alt='Favorite' title='Favorite' class='wpfp-img' />";
    }
}

function wpfp_clear_favorites() {
    if (wpfp_get_cookie()):
        foreach (wpfp_get_cookie() as $post_id => $val) {
            wpfp_set_cookie($post_id, "");
            wpfp_update_post_meta($post_id, -1);
        }
    endif;
    if (is_user_logged_in()) {
        $favorite_post_ids = wpfp_get_user_meta();
        if ($favorite_post_ids):
            foreach ($favorite_post_ids as $post_id) {
                wpfp_update_post_meta($post_id, -1);
            }
        endif;
        if (!delete_usermeta(wpfp_get_user_id(), WPFP_META_KEY)) {
            return false;
        }
    }
    return true;
}
function wpfp_do_remove_favorite($post_id) {
    if (!wpfp_check_favorited($post_id))
        return true;

    $a = true;
    if (is_user_logged_in()) {
        $user_favorites = wpfp_get_user_meta();
        $user_favorites = array_diff($user_favorites, array($post_id));
        $user_favorites = array_values($user_favorites);
        $a = wpfp_update_user_meta($user_favorites);
    }
    if ($a) $a = wpfp_set_cookie($_REQUEST['postid'], "");
    return $a;
}

function wpfp_content_filter($content) {
    if (is_page()):
        if (strpos($content,'{{wp-favorite-posts}}')!== false) {
            return str_replace('{{wp-favorite-posts}}', wpfp_list_favorite_posts(), $content);
        }
    endif;
    #if (is_single()):
        if (strpos($content,'[wpfp-link]')!== false) {
            return str_replace('[wpfp-link]', wpfp_link(1), $content);
        }
    #endif;
    return $content;
}
add_filter('the_content','wpfp_content_filter',7);
function wpfp_shortcode_func() {
    wpfp_list_favorite_posts();
}
add_shortcode('wp-favorite-posts', 'wpfp_shortcode_func');


function wpfp_add_js_script() {
	if (!wpfp_get_option('dont_load_js_file'))
		wp_enqueue_script( "wp-favroite-posts", WPFP_PATH . "/wpfp.js", array( 'jquery' ) );
}
add_action('wp_print_scripts', 'wpfp_add_js_script');

function wpfp_wp_print_styles() {
	if (!wpfp_get_option('dont_load_css_file'))
		echo "<link rel='stylesheet' id='wpfp-css' href='" . WPFP_PATH . "/wpfp.css' type='text/css' />" . "\n";
}
add_action('wp_print_styles', 'wpfp_wp_print_styles');

function wpfp_init() {
    $wpfp_options = array();
    $wpfp_options['add_favorite'] = "Add to favorites";
    $wpfp_options['added'] = "Added to favorites!";
    $wpfp_options['remove_favorite'] = "Remove from favorites";
    $wpfp_options['removed'] = "Removed from favorites!";
    $wpfp_options['clear'] = "Clear favorites";
    $wpfp_options['cleared'] = "<p>Favorites cleared!</p>";
    $wpfp_options['favorites_empty'] = "Favorite list is empty.";
    $wpfp_options['cookie_warning'] = "Your favorite posts saved to your browsers cookies. If you clear cookies also favorite posts will be deleted.";
    $wpfp_options['rem'] = "remove";
    $wpfp_options['text_only_registered'] = "Only registered users can favorite!";
    $wpfp_options['statics'] = 1;
    $wpfp_options['widget_title'] = '';
    $wpfp_options['widget_limit'] = 5;
    $wpfp_options['before_image'] = 'star.png';
    $wpfp_options['custom_before_image'] = '';
    $wpfp_options['dont_load_js_file'] = 0;
    $wpfp_options['dont_load_css_file'] = 0;
    add_option('wpfp_options', $wpfp_options, 'Favorite Posts Options');
}
add_action('activate_wp-favorite-posts/wp-favorite-posts.php', 'wpfp_init');

function wpfp_config() { include('wpfp-admin.php'); }

function wpfp_config_page() {
    if ( function_exists('add_submenu_page') )
        add_options_page(__('Favorite Posts'), __('Favorite Posts'), 'manage_options', 'wp-favorite-posts', 'wpfp_config');
}
add_action('admin_menu', 'wpfp_config_page');

function wpfp_update_user_meta($arr) {
    return update_usermeta(wpfp_get_user_id(),WPFP_META_KEY,$arr);
}

function wpfp_update_post_meta($post_id, $val) {
    $val = wpfp_get_post_meta($post_id) + $val;
    return add_post_meta($post_id, WPFP_META_KEY, $val, true) or update_post_meta($post_id, WPFP_META_KEY, $val);
}

function wpfp_delete_post_meta($post_id) {
    return delete_post_meta($post_id, WPFP_META_KEY);
}

function wpfp_get_cookie() {
    return $_COOKIE[WPFP_COOKIE_KEY];
}

function wpfp_get_options() {
   return get_option('wpfp_options');
}

function wpfp_get_user_id() {
    global $current_user;
    get_currentuserinfo();
    return $current_user->ID;
}

function wpfp_get_user_meta($user = "") {
    if (!empty($user)):
        $userdata = get_userdatabylogin($user);
        $user_id = $userdata->ID;
        return get_usermeta($user_id, WPFP_META_KEY);
    else:
        return get_usermeta(wpfp_get_user_id(), WPFP_META_KEY);
    endif;
}

function wpfp_get_post_meta($post_id) {
    $val = get_post_meta($post_id, WPFP_META_KEY, true);
    if ($val < 0) $val = 0;
    return $val;
}

function wpfp_set_cookie($post_id, $str) {
    $expire = time()+60*60*24*30;
    return setcookie("wp-favorite-posts[$post_id]", $str, $expire, "/");
}
function wpfp_is_user_favlist_public($user) {
    $user_opts = wpfp_get_user_options($user);
    if ($user_opts['list_is_public'])
        return true;
    else
        return false;
}

function wpfp_get_user_options($user) {
    $userdata = get_userdatabylogin($user);
    $user_id = $userdata->ID;
    return get_usermeta($user_id, WPFP_USER_OPTION_KEY);
}
function wpfp_is_user_can_edit() {
    if ($_REQUEST['user'])
        return false;
    return true;
}
function wpfp_remove_favorite_link($post_id) {
    if (wpfp_is_user_can_edit()) {
        $wpfp_options = wpfp_get_options();
        $class = 'wpfp-link remove-parent';
        $link = "<a id='rem_$post_id' class='$class' href='?wpfpaction=remove&amp;page=1&amp;postid=". $post_id ."' title='".wpfp_get_option('rem')."' rel='nofollow'>".wpfp_get_option('rem')."</a>";
        $link = apply_filters( 'wpfp_remove_favorite_link', $link );
        echo $link;
    }
}
function wpfp_clear_list_link() {
    if (wpfp_is_user_can_edit()) {
        $wpfp_options = wpfp_get_options();
        echo wpfp_before_link_img();
        echo wpfp_loading_img();
        echo "<a class='wpfp-link' href='?wpfpaction=clear' rel='nofollow'>". wpfp_get_option('clear') . "</a>";
    }
}
function wpfp_cookie_warning() {
    if (!is_user_logged_in() && !isset($_GET['user']) ):
        echo "<p>".wpfp_get_option('cookie_warning')."</p>";
    endif;
}

function wpfp_get_option($opt) {
    $wpfp_options = wpfp_get_options();
    return htmlspecialchars_decode( stripslashes ( $wpfp_options[$opt] ) );
}
